File: modules/printer-helper/android/src/main/java/church/b1/checkin/printer/BrotherProvider.kt
package church.b1.checkin.printer

import android.content.Context
import android.graphics.Bitmap
import android.util.Log

import com.brother.ptouch.sdk.NetPrinter
import com.brother.ptouch.sdk.Printer
import com.brother.sdk.lmprinter.Channel
import com.brother.sdk.lmprinter.OpenChannelError
import com.brother.sdk.lmprinter.PrintError
import com.brother.sdk.lmprinter.PrinterDriverGenerator
import com.brother.sdk.lmprinter.PrinterModel
import com.brother.sdk.lmprinter.setting.PrintImageSettings
import com.brother.sdk.lmprinter.setting.QLPrintSettings

class BrotherProvider {
    var context: Context? = null
    var readyToPrint = false
    var printerIP = ""
    var printerModel = "QL-1110NWB"

    var onError: ((source: String, message: String) -> Unit)? = null
    var onEvent: ((eventType: String, source: String, message: String) -> Unit)? = null

    fun scan(): Array<String> {
        val result = mutableListOf<String>()
        val printers = Printer()
        val models = arrayOf(
            "QL-1100", "QL-1110NWB", "QL-580N", "QL-710W",
            "QL-720NW", "QL-800", "QL-810W", "QL-820NWB", "QL-1115NWB"
        )
        val printerList = printers.getNetPrinters(models)

        for (printer in printerList) {
            result.add("${printer.modelName}~${printer.ipAddress}")
        }

        onEvent?.invoke("Scan", "BrotherProvider.kt", "Scan - ${result.size}")
        return result.toTypedArray()
    }

    fun checkInit(c: Context, ip: String, model: String) {
        printerIP = ip
        printerModel = model
        context = c
        onEvent?.invoke("Model Selected", "BrotherProvider.kt", "Printer Model - $printerModel")
    }

    fun configure() {
        // Reserved for future configuration
    }

    private fun getPrinterSettings(): QLPrintSettings {
        val settings = when (printerModel) {
            "Brother QL-1100" -> QLPrintSettings(PrinterModel.QL_1100)
            "Brother QL-580N" -> QLPrintSettings(PrinterModel.QL_580N)
            "Brother QL-710W" -> QLPrintSettings(PrinterModel.QL_710W)
            "Brother QL-720NW" -> QLPrintSettings(PrinterModel.QL_720NW)
            "Brother QL-800" -> QLPrintSettings(PrinterModel.QL_800)
            "Brother QL-810W" -> QLPrintSettings(PrinterModel.QL_810W)
            "Brother QL-820NWB" -> QLPrintSettings(PrinterModel.QL_820NWB)
            "Brother QL-1115NWB" -> QLPrintSettings(PrinterModel.QL_1115NWB)
            else -> QLPrintSettings(PrinterModel.QL_1110NWB)
        }
        return settings
    }

    fun printBitmaps(bmps: List<Bitmap>) {
        val channel = Channel.newWifiChannel(printerIP)
        val result = PrinterDriverGenerator.openChannel(channel)

        if (result.error.code != OpenChannelError.ErrorCode.NoError) {
            onError?.invoke("BrotherProvider.kt", "Error - Open Channel: ${result.error.code}")
            return
        }

        val dir = context?.getExternalFilesDir(null)
        val printerDriver = result.driver
        val printSettings = getPrinterSettings()

        printSettings.setPrintOrientation(PrintImageSettings.Orientation.Landscape)
        printSettings.setLabelSize(QLPrintSettings.LabelSize.DieCutW29H90)
        printSettings.setAutoCut(true)
        printSettings.setWorkPath(dir.toString())

        for (bmp in bmps) {
            val printError = printerDriver.printImage(bmp, printSettings)
            if (printError.code != PrintError.ErrorCode.NoError) {
                onError?.invoke("BrotherProvider.kt", "Error - Print Image: ${printError.code} - ${printError.errorDescription}")
            } else {
                onEvent?.invoke("Print", "BrotherProvider.kt", "Success - Print Image")
            }
        }

        printerDriver.closeChannel()
    }
}

File: modules/printer-helper/ios/src/BrotherProvider.swift
import Foundation
import UIKit
import BRLMPrinterKit

class BrotherProvider: PrintProviderProtocol {
    let brand = "Brother"
    var printerIP = ""
    var printerModel = "QL-1110NWB"
    var status = "Pending init"

    var onError: ((_ source: String, _ message: String) -> Void)?
    var onEvent: ((_ eventType: String, _ source: String, _ message: String) -> Void)?

    func scan() -> [String] {
        var result: [String] = []
        let option = BRLMNetworkSearchOption()
        option.searchDuration = 15
        option.printerList = [
            "Brother QL-1100", "Brother QL-1110NWB", "Brother QL-580N",
            "Brother QL-710W", "Brother QL-720NW", "Brother QL-800",
            "Brother QL-810W", "Brother QL-820NWB", "Brother QL-1115NWB"
        ]

        _ = BRLMPrinterSearcher.startNetworkSearch(option) { channel in
            let modelName = channel.extraInfo?.value(forKey: BRLMChannelExtraInfoKeyModelName) as? String ?? ""
            let ipAddress = channel.channelInfo
            result.append("\(modelName)~\(ipAddress)")
        }

        onEvent?("Scan", "BrotherProvider.swift", "Scan - \(result.count)")
        return result
    }

    func checkInit(ip: String, model: String) {
        printerIP = ip
        printerModel = model

        if !printerIP.isEmpty {
            status = printerIP
        } else {
            status = "No Printer"
        }

        onEvent?("Model Selected", "BrotherProvider.swift", "Printer Model - \(printerModel)")
    }

    func configure() {
        // Reserved for future configuration
    }

    private func getPrinterModel() -> BRLMPrinterModel {
        switch printerModel {
        // The bundled BRLMPrinterKit.xcframework does not define enum cases for
        // the QL-1100, QL-580N, or QL-800 models, so each is mapped to its
        // same-hardware-family sibling. These share DK label media handling, so
        // the default print settings are equivalent for label printing:
        //   QL-1100  -> QL-1110NWB (same hardware; 1110NWB adds network/BT)
        //   QL-580N  -> QL-720NW   (700-series successor; network variant)
        //   QL-800   -> QL-810W    (same 800 family; 810W adds WiFi)
        case "Brother QL-1100": return .QL_1110NWB
        case "Brother QL-1110NWB": return .QL_1110NWB
        case "Brother QL-580N": return .QL_720NW
        case "Brother QL-710W": return .QL_710W
        case "Brother QL-720NW": return .QL_720NW
        case "Brother QL-800": return .QL_810W
        case "Brother QL-810W": return .QL_810W
        case "Brother QL-820NWB": return .QL_820NWB
        case "Brother QL-1115NWB": return .QL_1115NWB
        default: return .QL_1110NWB
        }
    }

    func printImages(_ imageURIs: [String]) {
        let log: (String) -> Void = { [weak self] msg in
            self?.onEvent?("Debug", "BrotherProvider.swift", msg)
        }

        log("printImages start — ip=\(printerIP), model=\(printerModel), labels=\(imageURIs.count)")

        let channel = BRLMChannel(wifiIPAddress: printerIP)
        let generateResult = BRLMPrinterDriverGenerator.open(channel)

        guard generateResult.error.code == .noError,
              let printerDriver = generateResult.driver else {
            onError?("BrotherProvider.swift", "Open Channel failed: \(generateResult.error.code) (raw \(generateResult.error.code.rawValue)) for ip=\(printerIP)")
            return
        }
        log("Channel opened to \(printerIP)")

        defer {
            printerDriver.closeChannel()
        }

        guard let printSettings = BRLMQLPrintSettings(defaultPrintSettingsWith: getPrinterModel()) else {
            onError?("BrotherProvider.swift", "Could not create print settings for model: \(printerModel)")
            return
        }

        printSettings.labelSize = .dieCutW29H90
        printSettings.autoCut = true
        printSettings.printOrientation = .landscape
        log("Settings ready — sdkModel=\(getPrinterModel().rawValue), labelSize=dieCutW29H90, autoCut=true, orientation=landscape")

        for (index, uriString) in imageURIs.enumerated() {
            // react-native-view-shot returns a bare filesystem path on iOS (no
            // "file://" scheme), so build a proper file URL. URL(string:) on a
            // schemeless path produces a non-file URL the Brother SDK can't read.
            let path = uriString.hasPrefix("file://")
                ? (URL(string: uriString)?.path ?? uriString)
                : uriString
            let url = URL(fileURLWithPath: path)

            let exists = FileManager.default.fileExists(atPath: path)
            let attrs = try? FileManager.default.attributesOfItem(atPath: path)
            let size = (attrs?[.size] as? Int).map { "\($0) bytes" } ?? "n/a"
            log("Label \(index + 1)/\(imageURIs.count): uri=\(uriString)")
            log("  -> path=\(path) exists=\(exists) size=\(size)")
            log("  -> fileURL=\(url.absoluteString)")

            if !exists {
                onError?("BrotherProvider.swift", "Label \(index + 1): file not found at \(path)")
                continue
            }

            let printError = printerDriver.printImage(with: url, settings: printSettings)

            if printError.code != .noError {
                onError?("BrotherProvider.swift", "Print failed for label \(index + 1): \(printError.code) (raw \(printError.code.rawValue))")
            } else {
                onEvent?("Print", "BrotherProvider.swift", "Success — label \(index + 1) sent to printer")
            }
        }

        log("printImages complete")
    }
}
